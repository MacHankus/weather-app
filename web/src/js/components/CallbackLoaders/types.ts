import React, { PropsWithChildren } from 'react'

export interface CallbackLoaderRequiredProps {
    callback : ()=>Promise<{}>,
    failureCallback: Function,
    children : React.ReactElement
}
export type CallbackLoaderInputState = {error: boolean, helperText: string}

export interface CallbackLoaderInputProps extends CallbackLoaderRequiredProps {
    afterTyping?: boolean ,
    callback: (...args: any[])=>Promise<CallbackLoaderInputState>
}