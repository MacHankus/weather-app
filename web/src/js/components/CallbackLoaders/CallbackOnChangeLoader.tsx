import React, { ChangeEvent, useRef, useState } from 'react'
import { CallbackLoaderInputProps, CallbackLoaderInputState } from './types'

export const CallbackOnChangeLoader = ({ children, afterTyping, callback, failureCallback }: CallbackLoaderInputProps): React.ReactElement => {

    const [resolve, setResolve] = useState<CallbackLoaderInputState>({ error: false, helperText: "" })
    const [inputValue, setInputValue] = useState("")
    const inputValueRef = useRef(inputValue);
    inputValueRef.current = inputValue;

    const runCallbacks = async () => {
        try {
            const data = await callback(inputValueRef.current)
            setResolve(data)
        } catch (err) {
            failureCallback()
        }
    }

    const onChange = (e: ChangeEvent<HTMLInputElement>) => {
        const value = e.currentTarget.value
        setInputValue(value)
        if (afterTyping) {
            let timeToCheck = 500
            setTimeout(() => {
                if (value === inputValueRef.current) {
                    runCallbacks()
                }
            }, timeToCheck)
        }else{
            runCallbacks()
        }
    }

    return <>
        {
            React.cloneElement(
                React.Children.only(
                    children
                ),
                {
                    onChange,
                    value:inputValue,
                    ...resolve
                    
                }
            )
        }
    </>
}