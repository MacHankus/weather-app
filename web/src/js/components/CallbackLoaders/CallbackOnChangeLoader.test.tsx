import React from 'react'
import { CallbackOnChangeLoader } from './CallbackOnChangeLoader'
import { CallbackLoaderInputState } from './types'
import { render, fireEvent, waitFor, screen, findByTestId, getByText, prettyDOM, act, cleanup } from '@testing-library/react'
import { TextField } from '@material-ui/core'
import { isEmptyBindingElement } from 'typescript'
import userEvent from '@testing-library/user-event'
import "@testing-library/jest-dom/extend-expect"

describe("CallbackInputLoader simple input render.", () => {
  beforeEach(() => {
    cleanup
  })
  it("Match snapshot with initial render.", () => {
    const successCallback = () => {
      return new Promise<CallbackLoaderInputState>((res, rej) => {
        res({ error: false, helperText: "" })
      })
    }
    const failureCallback = () => {
      return new Promise((res, rej) => {
        console.log("Should never get here")
      })
    }
    const { container } = render(<CallbackOnChangeLoader callback={successCallback} failureCallback={failureCallback}>
      <TextField id="id-input" />
    </CallbackOnChangeLoader>)
    const rElement = container.firstChild
    expect(rElement).toMatchSnapshot()
  })
  it("With callback called should return no error on input.", async () => {

    const successCallback = () => {
      return new Promise<CallbackLoaderInputState>((res, rej) => {
        res({ error: false, helperText: "" })
      })
    }
    const failureCallback = () => {
      return new Promise((res, rej) => {
        console.log("Should never get here")
      })
    }

     const { container } =  render(<CallbackOnChangeLoader callback={successCallback} failureCallback={failureCallback}>
        <TextField id="id-input" />
      </CallbackOnChangeLoader>)



    const input = container.querySelector('#id-input')


    expect(input).not.toBeNull()

    if (input === null) {
      throw Error("input is null")
    }
    act(() => {
      userEvent.type(input, 'Username')
    })

    await waitFor(()=>{
      const helperText = container.querySelector('#id-input-helper-text')
      expect(helperText).toBeNull()
    })


  })
  it("Helper text appears.", async () => {
    jest.setTimeout(4000)
    const errorText = "This is error text."
    const Input = () => <TextField id="id-input" />
    const successCallback = () => {
      return new Promise<CallbackLoaderInputState>((res, rej) => {
        res({ error: true, helperText: errorText })
      })
    }
    const failureCallback = () => {
      return new Promise((res, rej) => {
        console.log("Should never get here")
      })
    }
    var container: Element | DocumentFragment
    const result = render(<CallbackOnChangeLoader callback={successCallback} failureCallback={failureCallback}>
      <Input />
    </CallbackOnChangeLoader>)
    container = result.container



    const input = container.querySelector('#id-input') as HTMLInputElement

    expect(input).not.toBeNull()

    if (input === null) {
      fail("input is null")
    }

    const typeToInput = "UsernameFailed"

    userEvent.type(input, typeToInput)

    expect(input.value).toEqual(typeToInput)

    function checkValueAfterTime() {
      return new Promise((res) => {
        setTimeout(() => {

          const helperText = container.querySelector('#id-input-text-helper')
          if (helperText === null) {
            fail('helperText is null.')
          }

          expect(helperText).toBeInTheDocument()
          res("")
        }, 1000)
      })
    }
    await checkValueAfterTime()

  })

  it("Succes callback called once .", async () => {
    jest.setTimeout(4000)
    const errorText = "This is error text."
    const Input = () => <TextField id="id-input" />
    const mockSuccessCallback = jest.fn(() => {
      return new Promise<CallbackLoaderInputState>((res, rej) => {
        res({ error: false, helperText: errorText })
      })
    })
    const mockFailureCallback = jest.fn(() => {
      return new Promise((res, rej) => {
        console.log("Should never get here")
      })
    })
    const { container } = render(<CallbackOnChangeLoader callback={mockSuccessCallback} failureCallback={mockFailureCallback}>
      <Input />
    </CallbackOnChangeLoader>)

    const input = container.querySelector('#id-input') as HTMLInputElement

    expect(input).not.toBeNull()

    function checkForCallbacksToBeZero() {
      return new Promise((res) => {
        setTimeout(() => {
          expect(mockSuccessCallback.mock.calls.length).toBe(0)
          console.log(mockSuccessCallback.mock.calls.length)
          expect(mockFailureCallback.mock.calls.length).toBe(0)
          res("")
        }, 500)
      })
    }

    await checkForCallbacksToBeZero()

    act(() => { userEvent.type(input, 'Input-Value') })


    function checkForCallbacksToBeOneZero() {
      return new Promise((res) => {
        setTimeout(() => {
          expect(mockSuccessCallback.mock.calls.length).toBe(1)
          console.log(mockSuccessCallback.mock.calls.length)
          expect(mockFailureCallback.mock.calls.length).toBe(0)
          res("")
        }, 500)
      })
    }

    //await checkForCallbacksToBeOneZero()
    await waitFor(()=>{
      expect(mockSuccessCallback.mock.calls.length).toBe(1)
      console.log(mockSuccessCallback.mock.calls.length)
      expect(mockFailureCallback.mock.calls.length).toBe(0)
    })
  })
})