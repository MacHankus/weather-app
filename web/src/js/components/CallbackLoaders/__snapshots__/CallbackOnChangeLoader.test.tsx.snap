// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CallbackInputLoader simple input render. Match snapshot with initial render. 1`] = `
<div
  class="MuiFormControl-root MuiTextField-root"
>
  <div
    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
  >
    <input
      aria-invalid="false"
      class="MuiInputBase-input MuiInput-input"
      id="id-input"
      type="text"
      value=""
    />
  </div>
  
</div>
`;
